### PR TITLE
Memory widget on the workbench main-window

### LINKED ISSUE #30757: Workbench: Display Overall Memory Usage
**Original Reporter**: ISIS/Diego Alba-Venero and Refl.

On Linux systems the [Out Of Memory Process Killer](https://docs.memset.com/other/linux-s-oom-process-killer) will take out processes using high amounts of memory when the available system becomes low. This is being more frequently seen on the IDAaaS systems at ISIS with workflows like SANS when too many reductions are ran, the available memory becomes low and the process just dies with no warning. It is quite often not clear that this has happened without using `dmesg` and finding a corresponding entry.

As a user I would like to see the amount of available system memory displayed in Workbench such that I can spot when resources are running low.

This task is to first mock up where this display would sit in the application. Once agreed it should be implemented as a `mantidqt` widget that could be reused by other applications and then incorporated into the appropriate location. This task should concern itself with a minimal working product of simply displaying the available system memory. Further enhancements, such as warnings, can be considered for the future.

### PR DESCRIPTION
**Pull request by srikanthravipati**
**Description of work.**

Added memory widget to the workbench main-window, to display the total memory usage

Report to: ISIS/Diego Alba-Venero and Refl

**To test:**

-> Pull and Checkout  branch "30757_memory_bar"

-> **On Windows**
    -> open "Run" app and type "%APPDATA%" and Enter
    -> In the revealed folder (.../AppData/Roaming/), go to "mantidproject" folder
    -> In the "mantidproject" folder, delete "mantidworkbench.ini" file

-> **On Mac**
    -> Remove "~/.config/mantidproject/mantidworkbench.ini" file

-> Open the workbench application, and memory widget should appear at the
     top right corner displaying memory usage info.

-> The memory_bar colour fill should be light-green, if the memory usage is sub-critical (<90%)
     or the colour fill should be red if the memory usage is critical (>= 90%)

**On Windows**
on Windows

**On Mac** (Thanks to @DanielMurphy22)
On Mac

Resolves #30757

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/applications/workbench/workbench/plugins/memorywidget/memoryinfo.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#
+# Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+#   NScD Oak Ridge National Laboratory, European Spallation Source,
+#   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+# SPDX - License - Identifier: GPL - 3.0 +
+#  This file is part of the mantidqt package
+#
+#
+from psutil import virtual_memory
+conversion_factor_to_GB = 1.0 / (1024 * 1024 * 1024)
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/plugins/memorywidget/memoryinfo.py:11]: Good to see a named constant. Can we plase have all capitals for module constants and a clear blank line to separate the imports from the constants?

File: qt/applications/workbench/workbench/plugins/memorywidget/memorypresenter.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+from qtpy.QtCore import QTimer
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/plugins/memorywidget/memorypresenter.py:1]: Can we add the license info here?

File: qt/applications/workbench/workbench/plugins/memorywidget/memorypresenter.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+from ..memorywidget.memoryinfo import get_memory_info
+
+
+class MemoryPresenter(object):
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/plugins/memorywidget/memorypresenter.py:6]: Can we add method docs to everything so we can signal to future developers what the intent is? An example of docs in Python is in (algorithm_factory_observer)[https://github.com/mantidproject/mantid/blob/master/qt/python/mantidqt/widgets/algorithmselector/algorithm_factory_observer.py]

File: qt/applications/workbench/workbench/plugins/memorywidget/memorypresenter.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        self.set_bar_color_at_start()
+        self.update_at_regular_intervals()
+
+    def update_at_regular_intervals(self):
+        self.timer.timeout.connect(self.update_memory_usage)
+        self.timer.start(10)
+
+    def set_bar_color_at_start(self):
+        current_value = self.view.memory_bar.value()
+        if current_value >= 90:
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/plugins/memorywidget/memorypresenter.py:20]: Similar, could we pull the `90` out as a named variable to aid in readability?

File: qt/applications/workbench/workbench/plugins/memorywidget/memoryview.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+def from_normal_to_critical(critical, current_value, new_value) -> bool:
+    return current_value < critical <= new_value
+
+
+def from_critical_to_normal(critical, current_value, new_value) -> bool:
+    return current_value >= critical > new_value
+
+
+class MemoryView(QWidget):
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/plugins/memorywidget/memoryview.py:34]: Similarly if we could add method docs here?

File: qt/applications/workbench/workbench/plugins/memorywidget/memoryview.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    def __init__(self, parent):
+        super(MemoryView, self).__init__(parent)
+
+        self.critical = 90
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/plugins/memorywidget/memoryview.py:38]: Is this `90` the same meaning as in the presenter? Can we share this by passing it in to view in the `__init__` method?

File: qt/applications/workbench/workbench/plugins/memorywidget/memoryview.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        else:
+            pass
+
+    def set_value(self, new_value, mem_used, mem_avail):
+        # newValue is the mem_used_percent(int)
+        current_value = self.memory_bar.value()
+        if current_value != new_value:
+            self.set_bar_color(current_value, new_value)
+            self.memory_bar.setValue(new_value)
+            display_str = "%3.1f" % mem_used + "/" + "%3.1f" % mem_avail + " GB " \
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/plugins/memorywidget/memoryview.py:56]: Would [`f-strings`](https://realpython.com/python-f-strings/#f-strings-a-new-and-improved-way-to-format-strings-in-python) improve the readability here?

**Review by martyngigg — CHANGES_REQUESTED** [review]
This is a really nice addition that I think users will really like. It will given them more visbility on what is happening with the memory on the system.

Could I ask though that we move the widget itself to the [`mantidqt`](https://github.com/mantidproject/mantid/tree/master/qt/python/mantidqt/widgets) package? The idea is that `mantidqt` is a package of reusable widgets that can be imported to other PyQt-based applications like workbench but others too, e.g. a standdalone mslice might choose to use it too. This means that `workbench` can know about `mantidqt` but not the other way around. There is then a thin wrapper in the workbench plugins directory much like some of the other widgets in there.

I've made some inline comments too but all fairly minor style things really. The MVP pattern has been done well and new tests have been added.

File: qt/applications/workbench/workbench/plugins/test/test_memorywidget/test_memorypresenter.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+#
+from workbench.plugins.memorywidget.memoryview import MemoryView, \
+    from_normal_to_critical, from_critical_to_normal
+from workbench.plugins.memorywidget.memorypresenter import MemoryPresenter
+
+import unittest
+from unittest import mock
+
+
+class MemoryPresenterTest(unittest.TestCase):
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/plugins/test/test_memorywidget/test_memorypresenter.py:18]: Could I ask the need for this test as well as the one in `mantidqt`? Is it not a copy?

Comment (srikanthravipati) [reply qt/applications/workbench/workbench/plugins/test/test_memorywidget/test_memorypresenter.py]: Thanks Martyn. I overlooked removing these tests from workbench.

**Review by martyngigg — APPROVED** [review]
This looks really nice now. Good code documentation and type hints used 👍🏻

File: qt/python/mantidqt/widgets/memorywidget/memoryinfo.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+#  This file is part of the mantidqt package
+#
+#
+from psutil import virtual_memory
+from typing import Tuple
+
+CONVERSION_FACTOR_BYTES_TO_GB = 1.0 / (1024 * 1024 * 1024)
+
+
+def get_memory_info() -> Tuple[int, float, float]:
```
[the remark this answers is not in the document]
Comment (srikanthravipati) [reply qt/python/mantidqt/widgets/memorywidget/memoryinfo.py:16]: Thanks Martyn

**Comment by thomashampson** [comment]
@peterfpeterson there is a small fix for a performance issue for this widget that you will want to pick up for the ornl-next branch. See pull request #31132

**Pull request merged**